### PR TITLE
Micronaut Worker Feature

### PR DESCRIPTION
**Pull request by musketyr**
Adds new [Micronaut Worker ](https://agorapulse.github.io/micronaut-worker/) feature.

Adds two example jobs - one simple and other distributed. The tests are generated for every language and framework supported. Java and Kotlin tests uses [Awaitility](https://github.com/awaitility/awaitility) for asynchronous testing.

### CODE REVIEW

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        return "Advanced distributed scheduling for Micronaut";
+    }
+
+    @Override
+    public boolean supports(ApplicationType applicationType) {
+        return applicationType == ApplicationType.DEFAULT;
+    }
+
+    @Override
+    public String getCategory() {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java:88]: I don't think other is a valid category.

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+    @Override
+    @NonNull
+    public String getName() {
+        return "micronaut-worker";
+    }
+
+    @Override
+    public String getTitle() {
+        return "Advanced distributed scheduling for Micronaut";
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java]: replace this with "Micronaut Worker"

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobGroovy.rocker.raw [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import org.reactivestreams.Publisher
+
+@@Singleton
+@@CompileStatic
+class EmailDigestDistributedJob {
+
+    final List<String> emailsSent = []
+
+    @@Cron('45 6 * * *')
+    Publisher<String> generateEmailsForDigest() {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobGroovy.rocker.raw]: Annotate  return with `@SingleResult`

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobGroovy.rocker.raw [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        return Publishers.just('user@@example.com')
+    }
+
+    @@FixedRate('1s')
+    void sendEmail(String email) {
+        emailsSent << email
+        println "Sending email to $email"
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobGroovy.rocker.raw]: remove `println` or replace it with log statement.

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobJava.rocker.raw [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    @@Cron("45 6 * * *")
+    public Publisher<String> generateEmailsForDigest() {
+        return Publishers.just("user@@example.com");
+    }
+
+    @@FixedRate("1s")
+    public void sendEmail(String email) {
+        emailsSent.add(email);
+        System.out.println("Sending email to " + email);
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobJava.rocker.raw]: remove `println` or replace it with log statement.

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobKotlin.rocker.raw [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    @@Cron("45 6 * * *")
+    fun generateEmailsForDigest(): Publisher<String> {
+        return Publishers.just("user@@example.com");
+    }
+
+    @@FixedRate("1s")
+    fun sendEmail(email: String) {
+        emailsSent.add(email);
+        println("Sending email to $email");
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobKotlin.rocker.raw]: remove `println` or replace it with log statement.

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobJava.rocker.raw [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+@@Singleton
+public class EmailDigestDistributedJob {
+
+    private final List<String> emailsSent = new ArrayList<>();
+
+    @@Cron("45 6 * * *")
+    public Publisher<String> generateEmailsForDigest() {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobJava.rocker.raw]: Annotate  return with `@SingleResult`

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobKotlin.rocker.raw [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import jakarta.inject.Singleton
+import org.reactivestreams.Publisher
+
+@@Singleton
+class EmailDigestDistributedJob {
+
+    internal val emailsSent: MutableList<String> = arrayListOf()
+
+    @@Cron("45 6 * * *")
+    fun generateEmailsForDigest(): Publisher<String> {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobKotlin.rocker.raw]: Annotate  return with `@SingleResult`

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobSpecGroovy.rocker.raw [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+class EmailDigestDistributedJobSpec extends Specification {
+
+    PollingConditions conditions = new PollingConditions()
+
+    @@Inject EmailDigestDistributedJob job
+    @@Inject JobManager jobManager
+
+    void 'send email using distributed job'() {
+        expect:
+            job.emailsSent.size() == 0
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobSpecGroovy.rocker.raw]: remove indentation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobSpecGroovy.rocker.raw [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+        when:
+            jobManager.forceRun(EmailDigestDistributedJob, 'generateEmailsForDigest')
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobSpecGroovy.rocker.raw]: remove indentation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobSpecGroovy.rocker.raw [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+        then:
+            conditions.eventually {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobSpecGroovy.rocker.raw]: remove indentation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+        addDependency(generatorContext);
+        addExampleCode(generatorContext);
+    }
+
+    private void addDependency(GeneratorContext generatorContext) {
+        generatorContext.addDependency(Dependency.builder()
+                .lookupArtifactId(ARTIFACT_ID)
+                .compile());
+
+        if (generatorContext.getLanguage() == Language.JAVA || generatorContext.getLanguage() == Language.KOTLIN) {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java]: extract to extra `awaitility` feature.

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobGroovy.rocker.raw [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+@@Singleton
+@@CompileStatic
+class EmailDigestSimpleJob {
+
+    private int emailsSent
+
+    @@Cron('45 6 * * *')
+    void sendDigestEmail() {
+        emailsSent++
+        println 'Emails has been sent at 6:45'
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobGroovy.rocker.raw]: remove println statement or replace it with log

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobJava.rocker.raw [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+@@Singleton
+public class EmailDigestSimpleJob {
+
+    private int emailsSent;
+
+    @@Cron("45 6 * * *")
+    public void sendDigestEmail() {
+        emailsSent++;
+        System.out.println("Emails has been sent at 6:45");
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobJava.rocker.raw]: remove println statement or replace it with log

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobKotlin.rocker.raw [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+@@Singleton
+class EmailDigestSimpleJob {
+
+    internal var emailsSent = 0
+
+    @@Cron("45 6 * * *")
+    fun sendDigestEmail() {
+        emailsSent++;
+        println("Emails has been sent at 6:45");
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobKotlin.rocker.raw]: remove println statement or replace it with log

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+class EmailDigestSimpleJobSpec extends Specification {
+
+    PollingConditions conditions = new PollingConditions()
+
+    @@Inject EmailDigestSimpleJob job
+    @@Inject JobManager jobManager
+
+    void 'send email using simple job'() {
+            expect:
+                job.emailsSent == 0
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw]: remove indentation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+            when:
+                jobManager.forceRun(EmailDigestSimpleJob)
+
+            then:
+                conditions.eventually {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw]: remove indentation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+
+    @@Inject EmailDigestSimpleJob job
+    @@Inject JobManager jobManager
+
+    void 'send email using simple job'() {
+            expect:
+                job.emailsSent == 0
+
+            when:
+                jobManager.forceRun(EmailDigestSimpleJob)
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw]: remove indentation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobTestJava.rocker.raw [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+    @@Inject EmailDigestSimpleJob job;
+    @@Inject JobManager jobManager;
+
+    @@Test
+    void testSendEmailsDirect() {
+        assertEquals(0, job.getEmailsSent());
+
+        jobManager.forceRun(EmailDigestSimpleJob.class);
+
+        await().atMost(1, SECONDS).until(() -> job.getEmailsSent() == 1);
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobTestJava.rocker.raw]: rest email sent after at the end of the test

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobTestKotest.rocker.raw [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+import org.junit.jupiter.api.Assertions
+import java.util.concurrent.TimeUnit
+
+@@MicronautTest
+class EmailDigestSimpleJobTest(private val jobManager: JobManager, private val job: EmailDigestSimpleJob): StringSpec({
+
+    "send emails direct" {
+        Assertions.assertEquals(0, job.emailsSent)
+        jobManager.forceRun(EmailDigestSimpleJob::class.java)
+        Awaitility.await().atMost(1, TimeUnit.SECONDS).until { job.emailsSent == 1 }
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobTestKotest.rocker.raw]: rest email sent after at the end of the test

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobTestKotlin.rocker.raw [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    lateinit var job: EmailDigestSimpleJob
+
+    @@Inject
+    lateinit var jobManager: JobManager
+
+    @@Test
+    fun testSendEmailsDirect() {
+        Assertions.assertEquals(0, job.emailsSent)
+        jobManager.forceRun(EmailDigestSimpleJob::class.java)
+        Awaitility.await().atMost(1, TimeUnit.SECONDS).until { job.emailsSent == 1 }
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobTestKotlin.rocker.raw]: rest email sent after at the end of the test

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+import io.micronaut.starter.feature.Feature
+import io.micronaut.starter.options.BuildTool
+import io.micronaut.starter.options.Language
+import spock.lang.Unroll
+
+class WorkerSpec extends ApplicationContextSpec {
+
+    void "Worker Feature override Feature->getThirdPartyDocumentation"() {
+        given:
+            List<ApplicationType> supportedApplicationTypes = [ApplicationType.DEFAULT]
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+        when:
+            Optional<Feature> featureOptional = findFeatureByName('micronaut-worker')
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+
+        then:
+            featureOptional.isPresent()
+
+        when:
+            Feature feature = featureOptional.get()
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+
+    void "Worker Feature override Feature->getThirdPartyDocumentation"() {
+        given:
+            List<ApplicationType> supportedApplicationTypes = [ApplicationType.DEFAULT]
+
+        when:
+            Optional<Feature> featureOptional = findFeatureByName('micronaut-worker')
+
+        then:
+            featureOptional.isPresent()
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+
+        when:
+            Feature feature = featureOptional.get()
+
+        then:
+            feature.getThirdPartyDocumentation()
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+
+        and: 'is in the OTHER category'
+            feature.getCategory() == "Other"
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+
+        and:
+            feature.isCommunity()
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+
+        and: 'supports only a ApplicationType.DEFAULT'
+            for (ApplicationType type : (ApplicationType.values() - supportedApplicationTypes)) {
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+            }
+            for (ApplicationType type : supportedApplicationTypes) {
+                assert feature.supports(type)
+            }
+    }
+
+    @Unroll("#buildTool with feature micronaut-worker adds dependency #groupId:#artifactId for #language")
+    void "verify micronaut-worker feature dependencies"(Language language, BuildTool buildTool, String groupId, String artifactId) {
+        given:
+            List<String> features = ['micronaut-worker']
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+            String coordinate = "${groupId}:${artifactId}"
+
+        when:
+            String template = new BuildBuilder(beanContext, buildTool)
+                    .features(features)
+                    .language(language)
+                    .render()
+
+        then:
+            if (buildTool.isGradle()) {
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Unroll("#buildTool with feature micronaut-worker adds dependency #groupId:#artifactId for #language")
+    void "verify micronaut-worker feature dependencies"(Language language, BuildTool buildTool, String groupId, String artifactId) {
+        given:
+            List<String> features = ['micronaut-worker']
+            String coordinate = "${groupId}:${artifactId}"
+
+        when:
+            String template = new BuildBuilder(beanContext, buildTool)
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+        then:
+            if (buildTool.isGradle()) {
+                assert template.contains("$configuration(\"$coordinate")
+            } else if (buildTool == BuildTool.MAVEN) {
+                assert template.contains("<artifactId>$artifactId</artifactId>")
+                assert template.contains("<groupId>$groupId</groupId>")
+            }
+
+        where:
+            language        | buildTool                 | configuration         | groupId                   | artifactId
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/agorapulse/worker/WorkerSpec.groovy]: remove indentation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+
+            then:
+                conditions.eventually {
+                    job.emailsSent == 1
+                }
+    }
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw:41]: reset email sent count

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobTestKotlin.rocker.raw [hunk 35]
```diff
@@ -0,0 +1,41 @@
...
+    @@Inject
+    lateinit var jobManager: JobManager
+
+    @@Test
+    fun testSendEmailsDistributed() {
+        Assertions.assertEquals(0, job.emailsSent.size)
+
+        jobManager.forceRun(EmailDigestDistributedJob::class.java, "generateEmailsForDigest")
+
+        Awaitility.await().atMost(2, TimeUnit.SECONDS).until { job.emailsSent.size == 1 }
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobTestKotlin.rocker.raw]: reset email sent when test completes

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobTestKotest.rocker.raw [hunk 36]
```diff
@@ -0,0 +1,31 @@
...
+
+@@MicronautTest
+@@Property(name = "worker.jobs.email-digest-distributed-job-send-email.enabled", value = "true")
+@@Property(name = "worker.jobs.email-digest-distributed-job-send-email.fixed-rate", value = "1ms")
+class EmailDigestDistributedJobTest(private val jobManager: JobManager, private val job: EmailDigestDistributedJob): StringSpec({
+
+    "send emails distributed" {
+        Assertions.assertEquals(0, job.emailsSent.size)
+        jobManager.forceRun(EmailDigestDistributedJob::class.java, "generateEmailsForDigest")
+        Awaitility.await().atMost(1, TimeUnit.SECONDS).until { job.emailsSent.size == 1 }
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobTestKotest.rocker.raw]: reset email sent when test completes

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobTestJava.rocker.raw [hunk 37]
```diff
@@ -0,0 +1,39 @@
...
+    @@Inject EmailDigestDistributedJob job;
+    @@Inject JobManager jobManager;
+
+    @@Test
+    void testSendEmailsDistributed() {
+        assertEquals(0, job.getEmailsSent().size());
+
+        jobManager.forceRun(EmailDigestDistributedJob.class, "generateEmailsForDigest");
+
+        await().atMost(2, SECONDS).until(() -> job.getEmailsSent().size() == 1);
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobTestJava.rocker.raw]: reset email sent when test completes

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobSpecGroovy.rocker.raw [hunk 38]
```diff
@@ -0,0 +1,43 @@
...
+                job.emailsSent.size() == 1
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobSpecGroovy.rocker.raw]: reset email sent when test completes

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java (hunk 1 above)
Comment (musketyr) [reply starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java:88]: but which one is the best?

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobGroovy.rocker.raw (hunk 3 above)
Comment (musketyr) [reply starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobGroovy.rocker.raw]: the method is supposed to return multiple items but for illustration purposes it only returns one so the feature don't have to depend on any reactive implementation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobJava.rocker.raw (hunk 7 above)
Comment (musketyr) [reply starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobJava.rocker.raw]: the method is supposed to return multiple items but for illustration purposes it only returns one so the feature don't have to depend on any reactive implementation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobKotlin.rocker.raw (hunk 8 above)
Comment (musketyr) [reply starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobKotlin.rocker.raw]: the method is supposed to return multiple items but for illustration purposes it only returns one so the feature don't have to depend on any reactive implementation

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java (hunk 1 above)
Comment (sdelamo) [reply starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/Worker.java:88]: 🤷

**Comment by musketyr** [comment]
@sdelamo I have updated the tests to be more robust and extracted the awaitility feature

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw (hunk 34 above)
Comment (musketyr) [reply starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestSimpleJobSpecGroovy.rocker.raw:41]: replaced with mocking

File: starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobTestJava.rocker.raw (hunk 37 above)
Comment (musketyr) [reply starter-core/src/main/java/io/micronaut/starter/feature/agorapulse/worker/template/emailDigestDistributedJobTestJava.rocker.raw]: replaced with mocking

**Comment by musketyr** [comment]
@sdelamo I've replaced `Publisher` with `Iterable` for simplicity.

**Review by sdelamo — APPROVED** [review]

**Comment by sdelamo** [comment]
@timyates we are getting this: 

Error: R] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project foo: Resolution of annotationProcessorPath dependencies failed: For artifact {info.picocli:picocli-codegen:null:jar}: The version cannot be empty. -> [Help 1]
[495](https://github.com/micronaut-projects/micronaut-starter/runs/6510838823?check_suite_focus=true#step:5:496)

this seems to be the picocli -codegen version any clue?

**Comment by timyates** [comment]
I wonder if the maven plugin relies on a version as we had with grpc...  Investigating

**Comment by timyates** [comment]
Yeah, it's the same issue we were seeing with grpc

We now only define the picocli version in the picocli bom, so we need to pull it through (but it is better that how we used to do it)...  I believe it's basically an issue we'll see whenever we remove a version from core that is used by an annotation processor in maven

there's a PR in this repo to fix starter, and a PR in core/parent to fix it over there

**Comment by timyates** [comment]
Ok the fix should be in 3.5.x, so I've kicked off the tests again...  We may need to merge 3.5.x into this...  How did you do it before?

**Comment by timyates** [comment]
All green 🥒

Merging, thanks @musketyr!

**Pull request merged**